### PR TITLE
[pt2] convert `out` params in `register_meta`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #101344

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -434,15 +432,6 @@ def meta__linalg_eigh(
...
     vals = A.new_empty(shape, dtype=toRealValueType(A.dtype))
 
-    if eigenvalues is not None and eigenvectors is not None:
-        assert isinstance(eigenvalues, TensorLike)
-        assert isinstance(eigenvectors, TensorLike)
-        eigenvalues = _maybe_resize_out(eigenvalues, vals.shape)
-        eigenvectors = _maybe_resize_out(eigenvectors, vecs.shape)
-        _safe_copy_out(copy_from=vals, copy_to=eigenvalues)  # type: ignore[arg-type]
-        _safe_copy_out(copy_from=vecs, copy_to=eigenvectors)  # type: ignore[arg-type]
-        return eigenvalues, eigenvectors
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:444]: Was needed due to `test_dispatch_meta_outplace_linalg_eigvalsh_cpu_float64` failing, see: https://github.com/pytorch/pytorch/pull/100964#discussion_r1189208522

File: torch/_prims_common/wrappers.py [hunk 2]
```diff
@@ -224,6 +224,12 @@ def _fn(*args, out=None, **kwargs):
                     if k not in kwargs:
                         kwargs[k] = out_attr
 
+            # out_names can sometimes appear in kwargs. But Python metas are
+            # defined without these params (hence this wrapper). This removes
+            # out_names from kwargs and sets their values as out.
+            if any(x in kwargs for x in out_names):
+                out = tuple(kwargs.pop(x) for x in out_names)
+
```
Comment (nkaretnikov) [inline torch/_prims_common/wrappers.py:232]: - Fixes `test_dispatch_meta_outplace_linalg_eigvalsh_cpu_float64` (see the other comment)
- Partially fixes `test_dispatch_symbolic_meta_outplace_all_strides_msort_cuda_float32` (fully fixed in the next PR in the stack).

Haven't looked deep, but both of the mentioned tests are `CompositeImplicitAutograd` funcs. CPP impl ends up calling a Python meta via the dispatcher. So maybe something related to the way the dispatcher passes args around. Maybe this: https://github.com/pytorch/pytorch/pull/79941#discussion_r907818660

**Review by lezcano — COMMENTED** [review]
This is a simple and nice fix, but does not get the signature of the function right, which may be problematic later down the road.

The issue here is that C++ functions implement out by passing two different parameters, while on python we pass a tuple. Have a look at how this is implemented in `register_decomposition`. I think you'd need to implement something similar in `register_meta`.

**Comment by ezyang** [comment]
deferring to @lezcano on this

**Comment by nkaretnikov** [comment]
Moved the `out`-handling code from `decomposition_decorator` into a separate func, so I could reuse it.

**Review by lezcano — APPROVED** [review]
Easy peasy. Wasn't there another function that also had manual out handling due to this issue? If so, could you change it here? If not, this looks great.

**Comment by nkaretnikov** [comment]
Updated `meta_take` to use `out_wrapper` as well.

**Pull request closed without merging**